pyro/ops/gaussian.py

```diff
@@ -34,28 +35,147 @@ def __init__(self, log_normalizer, info_vec, precision):
     def dim(self):
         return self.info_vec.size(-1)
 
-    @property
+    @lazy_property
     def batch_shape(self):
         return broadcast_shape(self.log_normalizer.shape,
                                self.info_vec.shape[:-1],
                                self.precision.shape[:-2])
 
+    def expand(self, batch_shape):
+        n = self.info_vec.size(-1)
+        log_normalizer = self.log_normalizer.expand(batch_shape)
+        info_vec = self.info_vec.expand(batch_shape + (n,))
+        precision = self.precision.expand(batch_shape + (n, n))
```

Comment:
I was wondering if we need to expand `info_vec` and `precision` here. I think we can manipulate `reshape` of `info_vec` manually. But I can't find a good solution for this. Something likes
```
def find_unbroadcasted_new_shape(shape, unbroadcasted_shape, new_shape):
    # return expand_shape & unbroadcasted_new_shape
    # where we will expand `info_vec` to `expand_shape`
    # and then reshape it to `unbroadcasted_new_shape`
    # We'll do this work at at `Gaussian.reshape` operator
```
Btw, I would like to defer this to another PR, when we have an example to test out the performance. I just raise this to see what you think?

Reply:
For cost purposes I've intentionally separated `.reshape()` and `.expand()`, so `.reshape()` may fail if `.expand()` has not been performed first. Similarly, `__getitem__` assumes `.expand()` has been called. It might be possible to keep tensors smaller (as in `._unbroadcasted_*`), but doing so is complex and requires lots of Python manipulation. I'd prefer to defer this optimization until profiling shows that it is worth the added complexity.

Reply:
yup, I tried to implement this down this noon but it is a bit complicated to handle various patterns of expand and reshape. Will play more with it to see if I can come up with a simple solution.